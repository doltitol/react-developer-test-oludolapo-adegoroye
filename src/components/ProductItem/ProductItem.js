import React, { PureComponent } from 'react';
import { Icons } from '../../assets/images/Icons';
import { numberCommaFormatter } from '../../util';
import './productitem.style.scss';
import { Link } from 'react-router-dom';
import { client } from '../../config/apolloClient';
import { allResolvers } from '../../graphql/resolvers';
import { connect } from 'react-redux';
import { addToCart } from '../../redux/actions/cartActions';

export class ProductItem extends PureComponent {
  state = {
    showAddToCart: false,
    product: {},
  };
  async getProduct() {
    const id = this.props.productId;
    return await client.query({
      query: allResolvers.PRODUCT,
      variables: {
        productId: id,
      },
      fetchPolicy: 'network-only',
    });
  }
  componentDidMount() {
    this.getProduct().then((result) => {
      this.setState((prevState) => {
        return {
          product: prevState.product === result ? prevState.product : result,
        };
      });
    });
  }
  componentDidUpdate(prevProps) {
    if (this.props.productId !== prevProps.productId) {
      this.getProduct().then((result) =>
        this.setState((prevState) => {
          return {
            product: prevState.product === result ? prevState.product : result,
          };
        })
      );
    }
  }
  render() {
    const activeAttribute = this.state.product?.data
      ? this.state.product?.data.product.attributes.filter(
          (attr) => attr.id === 'Size' || attr.id === 'Capacity'
        ).length !== 0
        ? this.state.product?.data.product.attributes.filter(
            (attr) => attr.id === 'Size' || attr.id === 'Capacity'
          )[0].items[0].id
        : ''
      : null;
    const activeColor = this.state.product?.data
      ? this.state.product?.data.product.attributes.filter(
          (attr) => attr.id === 'Color'
        ).length !== 0
        ? this.state.product?.data.product.attributes.filter(
            (attr) => attr.id === 'Color'
          )[0].items[0].id
        : ''
      : null;
    return (
      <>
        {this.state.product?.data && (
          <div
            className='product-item'
            onMouseOver={() => this.setState({ showAddToCart: true })}
            onMouseLeave={() => this.setState({ showAddToCart: false })}
          >
            <Link to={`/product/${this.state.product.data.product.id}`}>
              <div className='product-item-image'>
                <img
                  src={this.state.product.data.product.gallery[0]}
                  alt={this.state.product.data.product.name}
                  className='product-item-image-img'
                />
                {!this.state.product.data.product.inStock && (
                  <div className='product-item-image-screen'>
                    <p className='product-item-image-screen-text'>
                      out of stock
                    </p>
                  </div>
                )}
              </div>
            </Link>
            {this.state.showAddToCart && (
              <Icons.AddToCart
                size={80}
                color={
                  !this.state.product.data.product.inStock
                    ? 'rgba(94, 206, 123, 0.8)'
                    : '#5ECE7B'
                }
                className='product-item-addtocart'
                onClick={() =>
                  this.state.product.data.product.inStock
                    ? this.props.addToCart(
                        this.state.product.data.product,
                        activeColor,
                        activeAttribute
                      )
                    : null
                }
              />
            )}
            <Link to={`/product/${this.state.product.data.product.id}`}>
              <p
                className='product-item-name'
                style={{
                  color: this.state.product.data.product.inStock
                    ? '#1D1F22'
                    : '#8D8F9A',
                }}
              >
                {this.state.product.data.product.brand} -{' '}
                {this.state.product.data.product.name}
              </p>
            </Link>
            {this.state.product.data.product.prices
              .filter((price) => price.currency.symbol === this.props.currency)
              .map((item) => (
                <p
                  className='product-item-price'
                  key={item.currency.symbol}
                  style={{
                    color: this.state.product.data.product.inStock
                      ? '#1D1F22'
                      : '#8D8F9A',
                  }}
                >
                  {item.currency.symbol} {numberCommaFormatter(item.amount)}
                </p>
              ))}
          </div>
        )}
      </>
    );
  }
}
const mapStateToProps = (state) => {
  return {
    currency: state.cart.currency,
  };
};
const mapDispatchToProps = (dispatch) => {
  return {
    addToCart: (product, activeColor, activeAttribute) => {
      dispatch(addToCart(product, activeColor, activeAttribute));
    },
  };
};
export default connect(mapStateToProps, mapDispatchToProps)(ProductItem);
