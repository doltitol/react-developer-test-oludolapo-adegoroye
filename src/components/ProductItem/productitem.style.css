@import url("https://fonts.googleapis.com/css2?family=Raleway:wght@400;600;700&family=Roboto+Condensed:wght@700&family=Roboto:wght@400;500&family=Source+Sans+Pro&display=swap");
* {
  -webkit-box-sizing: border-box;
          box-sizing: border-box;
  padding: 0;
  margin: 0;
  scroll-behavior: smooth;
}

body {
  background: #ffffff;
}

a {
  text-decoration: none;
}

.slide-right {
  -webkit-animation: slide-right 0.5s cubic-bezier(0.25, 0.46, 0.45, 0.94) both;
  animation: slide-right 0.5s cubic-bezier(0.25, 0.46, 0.45, 0.94) both;
}

@-webkit-keyframes slide-right {
  0% {
    -webkit-transform: translateX(50px);
    transform: translateX(50px);
  }
  100% {
    -webkit-transform: translateX(0px);
    transform: translateX(0px);
  }
}

@keyframes slide-right {
  0% {
    -webkit-transform: translateX(50px);
    transform: translateX(50px);
  }
  100% {
    -webkit-transform: translateX(0px);
    transform: translateX(0px);
  }
}

.slide-down {
  -webkit-animation: slide-down 0.5s cubic-bezier(0.25, 0.46, 0.45, 0.94) both;
  animation: slide-down 0.5s cubic-bezier(0.25, 0.46, 0.45, 0.94) both;
}

@-webkit-keyframes slide-down {
  0% {
    -webkit-transform: translateY(-25%);
    transform: translateY(-25%);
  }
  100% {
    -webkit-transform: translateY(0px);
    transform: translateY(0px);
  }
}

@keyframes slide-down {
  0% {
    -webkit-transform: translateY(-25%);
    transform: translateY(-25%);
  }
  100% {
    -webkit-transform: translateY(0px);
    transform: translateY(0px);
  }
}

.product-item {
  margin: 0px;
  padding: 20px;
  width: 100%;
  height: 100%;
  background-color: #ffffff;
  -webkit-transition: all 0.7s ease-in-out;
  transition: all 0.7s ease-in-out;
  position: relative;
  text-decoration: none;
}

.product-item:hover {
  -webkit-filter: drop-shadow(0px 4px 35px rgba(168, 172, 176, 0.19));
          filter: drop-shadow(0px 4px 35px rgba(168, 172, 176, 0.19));
  -webkit-box-shadow: 0px 4px 35px rgba(168, 172, 176, 0.19);
          box-shadow: 0px 4px 35px rgba(168, 172, 176, 0.19);
  cursor: pointer;
}

.product-item-image {
  position: relative;
  height: 400px;
}

@media screen and (max-width: 1600px) {
  .product-item-image {
    height: 320px;
  }
}

@media screen and (max-width: 1400px) {
  .product-item-image {
    height: 250px;
  }
}

@media screen and (max-width: 1200px) {
  .product-item-image {
    height: 320px;
  }
}

@media screen and (max-width: 960px) {
  .product-item-image {
    height: 270px;
  }
}

@media screen and (max-width: 550px) {
  .product-item-image {
    height: 220px;
  }
}

.product-item-image-img {
  width: 100%;
  height: 400px;
  -o-object-fit: fill;
     object-fit: fill;
}

@media screen and (max-width: 1600px) {
  .product-item-image-img {
    width: 100%;
    height: 320px;
  }
}

@media screen and (max-width: 1400px) {
  .product-item-image-img {
    width: 100%;
    height: 250px;
  }
}

@media screen and (max-width: 1200px) {
  .product-item-image-img {
    width: 100%;
    height: 320px;
  }
}

@media screen and (max-width: 960px) {
  .product-item-image-img {
    width: 100%;
    height: 270px;
  }
}

@media screen and (max-width: 550px) {
  .product-item-image-img {
    width: 100%;
    height: 220px;
  }
}

.product-item-image-screen {
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  height: 100%;
  background-color: rgba(255, 255, 255, 0.6);
  display: -webkit-box;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
      -ms-flex-pack: center;
          justify-content: center;
  -webkit-box-align: center;
      -ms-flex-align: center;
          align-items: center;
}

.product-item-image-screen-text {
  text-transform: uppercase;
  font-family: "Raleway", sans-serif;
  font-weight: 400;
  color: #1D1F22;
  font-size: 24px;
  line-height: 38.8px;
  color: #8D8F9A;
}

.product-item-name {
  font-family: "Raleway", sans-serif;
  font-weight: 300;
  color: #1D1F22;
  font-size: 18px;
  line-height: 28.8px;
  margin-top: 20px;
}

.product-item-price {
  font-family: "Raleway", sans-serif;
  font-weight: 500;
  color: #1D1F22;
  font-size: 18px;
  line-height: 28.8px;
}

.product-item-addtocart {
  position: absolute;
  right: 25px;
  bottom: 50px;
  -webkit-transition: all 0.7s ease-in-out;
  transition: all 0.7s ease-in-out;
}

@media screen and (max-width: 550px) {
  .product-item-addtocart {
    right: 25px;
    bottom: 60px;
  }
}
/*# sourceMappingURL=productitem.style.css.map */