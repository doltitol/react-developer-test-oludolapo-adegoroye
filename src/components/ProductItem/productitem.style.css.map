{
    "version": 3,
    "mappings": "ACAA,OAAO,CAAC,wKAAI;AAkBZ,AAAA,CAAC,CAAC;EACA,UAAU,EAAE,UAAU;EACtB,OAAO,EAAE,CAAC;EACV,MAAM,EAAE,CAAC;EACT,eAAe,EAAE,MAAM;CACxB;;AACD,AAAA,IAAI,CAAA;EACJ,UAAU,ECrBF,OAAO;CDsBd;;AACD,AAAA,CAAC,CAAA;EACD,eAAe,EAAE,IAAI;CACpB;;AACD,AAAA,YAAY,CAAC;EACX,iBAAiB,EAAE,WAAW,CAAC,IAAI,CAAC,oCAAoC,CAAC,IAAI;EAC7E,SAAS,EAAE,WAAW,CAAC,IAAI,CAAC,oCAAoC,CAAC,IAAI;CACtE;;AAED,kBAAkB,CAAlB,WAAkB;EAChB,EAAE;IACA,iBAAiB,EAAE,gBAAgB;IACnC,SAAS,EAAE,gBAAgB;;EAE7B,IAAI;IACF,iBAAiB,EAAE,eAAe;IAClC,SAAS,EAAE,eAAe;;;;AAI9B,UAAU,CAAV,WAAU;EACR,EAAE;IACA,iBAAiB,EAAE,gBAAgB;IACnC,SAAS,EAAE,gBAAgB;;EAE7B,IAAI;IACF,iBAAiB,EAAE,eAAe;IAClC,SAAS,EAAE,eAAe;;;;AAI9B,AAAA,WAAW,CAAC;EACV,iBAAiB,EAAE,UAAU,CAAC,IAAI,CAAC,oCAAoC,CAAC,IAAI;EAC5E,SAAS,EAAE,UAAU,CAAC,IAAI,CAAC,oCAAoC,CAAC,IAAI;CACrE;;AAED,kBAAkB,CAAlB,UAAkB;EAChB,EAAE;IACA,iBAAiB,EAAE,gBAAgB;IACnC,SAAS,EAAE,gBAAgB;;EAE7B,IAAI;IACF,iBAAiB,EAAE,eAAe;IAClC,SAAS,EAAE,eAAe;;;;AAI9B,UAAU,CAAV,UAAU;EACR,EAAE;IACA,iBAAiB,EAAE,gBAAgB;IACnC,SAAS,EAAE,gBAAgB;;EAE7B,IAAI;IACF,iBAAiB,EAAE,eAAe;IAClC,SAAS,EAAE,eAAe;;;;AD9E9B,AAAA,aAAa,CAAC;EACZ,MAAM,EAAE,GAAG;EACX,OAAO,EAAE,IAAI;EACb,KAAK,EAAE,IAAI;EACX,MAAM,EAAE,IAAI;EACZ,gBAAgB,EEHV,OAAO;EFIb,UAAU,EAAE,oBAAoB;EAChC,QAAQ,EAAE,QAAQ;EAClB,eAAe,EAAE,IAAI;CAqFtB;;AA7FD,AASE,aATW,AASV,MAAM,CAAC;EACN,MAAM,EAAE,mDAAmD;EAC3D,UAAU,EAAE,GAAG,CAAC,GAAG,CAAC,IAAI,CAAC,yBAAyB;EAClD,MAAM,EAAE,OAAO;CAChB;;AACA,AAAD,mBAAO,CAAC;EACN,QAAQ,EAAE,QAAQ;EAClB,MAAM,EAAE,KAAK;CA2Dd;;AA1DK,MAAM,CAAC,MAAM,MAAM,SAAS,EAAE,MAAM;EAHzC,AAAD,mBAAO,CAAC;IAIF,MAAM,EAAE,KAAK;GAyDlB;;;AAvDG,MAAM,CAAC,MAAM,MAAM,SAAS,EAAE,MAAM;EANvC,AAAD,mBAAO,CAAC;IAOF,MAAM,EAAE,KAAK;GAsDlB;;;AApDG,MAAM,CAAC,MAAM,MAAM,SAAS,EAAE,MAAM;EATvC,AAAD,mBAAO,CAAC;IAUF,MAAM,EAAE,KAAK;GAmDlB;;;AAjDG,MAAM,CAAC,MAAM,MAAM,SAAS,EAAE,KAAK;EAZtC,AAAD,mBAAO,CAAC;IAaF,MAAM,EAAE,KAAK;GAgDlB;;;AA7CG,MAAM,CAAC,MAAM,MAAM,SAAS,EAAE,KAAK;EAhBtC,AAAD,mBAAO,CAAC;IAiBF,MAAM,EAAE,KAAK;GA4ClB;;;AA1CE,AAAD,uBAAK,CAAC;EACJ,KAAK,EAAE,IAAI;EACX,MAAM,EAAE,KAAK;EACb,UAAU,EAAE,IAAI;CAsBjB;;AArBC,MAAM,CAAC,MAAM,MAAM,SAAS,EAAE,MAAM;EAJrC,AAAD,uBAAK,CAAC;IAKF,KAAK,EAAE,IAAI;IACX,MAAM,EAAE,KAAK;GAmBhB;;;AAjBC,MAAM,CAAC,MAAM,MAAM,SAAS,EAAE,MAAM;EARrC,AAAD,uBAAK,CAAC;IASF,KAAK,EAAE,IAAI;IACX,MAAM,EAAE,KAAK;GAehB;;;AAbC,MAAM,CAAC,MAAM,MAAM,SAAS,EAAE,MAAM;EAZrC,AAAD,uBAAK,CAAC;IAaF,KAAK,EAAE,IAAI;IACX,MAAM,EAAE,KAAK;GAWhB;;;AATC,MAAM,CAAC,MAAM,MAAM,SAAS,EAAE,KAAK;EAhBpC,AAAD,uBAAK,CAAC;IAiBF,KAAK,EAAE,IAAI;IACX,MAAM,EAAE,KAAK;GAOhB;;;AAJC,MAAM,CAAC,MAAM,MAAM,SAAS,EAAE,KAAK;EArBpC,AAAD,uBAAK,CAAC;IAsBF,KAAK,EAAE,IAAI;IACX,MAAM,EAAE,KAAK;GAEhB;;;AACA,AAAD,0BAAQ,CAAA;EACR,QAAQ,EAAE,QAAQ;EAClB,GAAG,EAAE,CAAC;EACN,IAAI,EAAE,CAAC;EACP,KAAK,EAAE,CAAC;EACR,MAAM,EAAE,IAAI;EACZ,gBAAgB,EAAE,wBAAqB;EACvC,OAAO,EAAE,IAAI;EACb,eAAe,EAAE,MAAM;EACvB,WAAW,EAAE,MAAM;CAMlB;;AALA,AAAD,+BAAM,CAAA;EACN,cAAc,EAAE,SAAS;EGhE3B,WAAW,EANC,SAAS,EAAE,UAAU;EAOjC,WAAW,EHgEsB,GAAG;EG/DpC,KAAK,EDRC,OAAO;ECSb,SAAS,EH8DU,IAAI;EG7DvB,WAAW,EH6Dc,MAAM;EAC7B,KAAK,EEpEF,OAAO;CFqET;;AAGF,AAAD,kBAAM,CAAC;EGtEP,WAAW,EANC,SAAS,EAAE,UAAU;EAOjC,WAAW,EHsEsB,GAAG;EGrEpC,KAAK,EDRC,OAAO;ECSb,SAAS,EHoEU,IAAI;EGnEvB,WAAW,EHmEc,MAAM;EAC7B,UAAU,EAAE,IAAI;CACjB;;AACA,AAAD,mBAAO,CAAC;EG1ER,WAAW,EANC,SAAS,EAAE,UAAU;EAOjC,WAAW,EH0EsB,GAAG;EGzEpC,KAAK,EDRC,OAAO;ECSb,SAAS,EHwEU,IAAI;EGvEvB,WAAW,EHuEc,MAAM;CAC9B;;AACA,AAAD,uBAAW,CAAC;EACV,QAAQ,EAAE,QAAQ;EAClB,KAAK,EAAE,IAAI;EACX,MAAM,EAAE,IAAI;EACZ,UAAU,EAAE,oBAAoB;CAKjC;;AAJE,MAAM,CAAC,MAAM,MAAM,SAAS,EAAE,KAAK;EALrC,AAAD,uBAAW,CAAC;IAMR,KAAK,EAAE,IAAI;IACb,MAAM,EAAE,IAAI;GAEb",
    "sources": [
        "productitem.style.scss",
        "../../styles/index.scss",
        "../../styles/_colors.scss",
        "../../styles/_fonts.scss",
        "../../styles/_colors.scss"
    ],
    "names": [],
    "file": "productitem.style.css"
}