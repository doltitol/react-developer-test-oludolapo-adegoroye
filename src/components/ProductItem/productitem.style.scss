@import '../../styles/index.scss';

.product-item {
  margin: 0px;
  padding: 20px;
  width: 100%;
  height: 100%;
  background-color: $white;
  transition: all 0.7s ease-in-out;
  position: relative;
  text-decoration: none;
  &:hover {
    filter: drop-shadow(0px 4px 35px rgba(168, 172, 176, 0.19));
    box-shadow: 0px 4px 35px rgba(168, 172, 176, 0.19);
    cursor: pointer;
  }
  &-image {
    position: relative;
    height: 400px;
        @media screen and (max-width: 1600px) {
        height: 320px;
      }
      @media screen and (max-width: 1400px) {
        height: 250px;
      }
      @media screen and (max-width: 1200px) {
        height: 320px;
      }
      @media screen and (max-width: 960px) {
        height: 270px;
      }

      @media screen and (max-width: $sm) {
        height: 220px;
      }
    &-img {
      width: 100%;
      height: 400px;
      object-fit: fill;
      @media screen and (max-width: 1600px) {
        width: 100%;
        height: 320px;
      }
      @media screen and (max-width: 1400px) {
        width: 100%;
        height: 250px;
      }
      @media screen and (max-width: 1200px) {
        width: 100%;
        height: 320px;
      }
      @media screen and (max-width: 960px) {
        width: 100%;
        height: 270px;
      }

      @media screen and (max-width: $sm) {
        width: 100%;
        height: 220px;
      }
    }
    &-screen{
    position: absolute;
    top: 0;
    left: 0;
    right: 0;
    height: 100%;
    background-color: rgba(255,255,255,0.6);
    display: flex;
    justify-content: center;
    align-items: center;
    &-text{
    text-transform: uppercase;
    @include raleway(24px, 38.8px, 400);
    color: $grey;
    }
    }
  }
  &-name {
    @include raleway(18px, 28.8px, 300);
    margin-top: 20px;
  }
  &-price {
    @include raleway(18px, 28.8px, 500);
  }
  &-addtocart {
    position: absolute;
    right: 25px;
    bottom: 50px;
    transition: all 0.7s ease-in-out;
     @media screen and (max-width: $sm) {
      right: 25px;
    bottom: 60px;
     }
  }
}
