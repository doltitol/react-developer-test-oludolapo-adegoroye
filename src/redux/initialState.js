export const initialState = {
  cartItems: JSON.parse(localStorage.getItem('cartItems') || '[]'),
  cartQuantity: JSON.parse(localStorage.getItem('cartItems') || '[]').reduce(
    (acc, item) => acc + item.qty,
    0
  ),
  currency: '$',
};
