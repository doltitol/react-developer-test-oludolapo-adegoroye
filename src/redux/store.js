import { createStore, combineReducers } from 'redux';
import cartReducer from './reducers/cartReducers';

const rootReducer = combineReducers({
  cart: cartReducer,
});
export const store = createStore(rootReducer);
