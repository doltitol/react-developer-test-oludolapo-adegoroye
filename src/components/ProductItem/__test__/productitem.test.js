import { render } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { Provider } from 'react-redux';
import { store } from '../../../redux/store';
import ProductItem from '../ProductItem';

describe('<ProductItem /> spec', () => {
  it('renders the empty component', () => {
    const container = render(
      <Provider store={store}>
        <ProductItem productId='apple-iphone-12-pro' />
      </Provider>
    );
    expect(container.children).toMatchSnapshot();
  });
});
