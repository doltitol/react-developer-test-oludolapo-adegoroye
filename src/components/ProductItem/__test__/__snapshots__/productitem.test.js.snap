// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductItem /> spec renders the empty component 1`] = `undefined`;
